### PR TITLE
Reduce the use of the preprocessor

### PR DESCRIPTION
**Pull request by khyperia**
Prefer to use runtime checks instead. Only use a the preprocessor where an API actually matters at compile-time.

Remaining preprocessor checks:

* `DesktopAnalyzerAssemblyLoader`/`CoreClrAnalyzerAssemblyLoader`/`ShadowCopyAnalyzerAssemblyLoader`: preprocessor needed for badly-structured inherentence for `ShadowCopy`, and for the API `AssemblyLoadContext`.
* `NamedPipeClientConnection.cs` `NamedPipeServerStream` constructor (the `WRITE_DAC` stuff)
* `BuildServerConnection.cs` `CheckIdentityUnix` - private reflection not available on netstandard (for authentication of pipes). (However, it doesn't matter, since the only netstandard project here is clientside, not serverside)
* `BuildServerConnection.cs` `GetPipeSecurity` - `PipeStream.GetAccessControl` not available on netstandard (preprocessor changes it to reflection).
* Bunches of tests.

### CODE REVIEW

**Comment by khyperia** [comment]
retest this please

File: src/Compilers/Server/VBCSCompilerTests/CompilerServerTests.cs [hunk 1]
```diff
@@ -198,11 +198,7 @@ private static (T Result, string Output) UseTextWriter<T>(Encoding encoding, Fun
             var client = ServerUtil.CreateBuildClient(language);
             client.TimeoutOverride = Timeout.Infinite;
 
-#if NET461
-            var sdkDir = RuntimeEnvironment.GetRuntimeDirectory();
-#else
-            string sdkDir = null;
-#endif
+            var sdkDir = ServerUtil.DefaultSdkDirectory;
```
Comment (AlekseyTs) [inline src/Compilers/Server/VBCSCompilerTests/CompilerServerTests.cs]: >ServerUtil.DefaultSdkDirectory [](start = 25, length = 30)

Why not BuildClient.GetSystemSdkDirectory?

File: src/Compilers/Shared/BuildClient.cs [hunk 2]
```diff
@@ -40,6 +40,18 @@ internal abstract class BuildClient
     {
         protected static bool IsRunningOnWindows => Path.DirectorySeparatorChar == '\\';
 
+        public static string GetSystemSdkDirectory()
```
Comment (AlekseyTs) [inline src/Compilers/Shared/BuildClient.cs]: >GetSystemSdkDirectory [](start = 29, length = 21)

Consider adding a doc comment

File: src/Compilers/Server/VBCSCompilerTests/CompilerServerTests.cs (hunk 1 above)
Comment (khyperia) [reply src/Compilers/Server/VBCSCompilerTests/CompilerServerTests.cs]: `ServerUtil.DefaultSdkDirectory` is already an existing API in tests, used elsewhere in test code. So, the reason is consistency.

Comment (AlekseyTs) [reply src/Compilers/Server/VBCSCompilerTests/CompilerServerTests.cs]: Perhaps we should rename ```ServerUtil.DefaultSdkDirectory``` to use consistent naming.

**Pull request merged**